Drop FOREACH_WORD

keszybz opened the pull request:
This came out longer than I expected...

mrc0mmand commented:
Something went wrong with legacy cgroups management (in the CentOS 7 job):

```
336/474 test-cgroup                             FAIL     0.28 s (killed by signal 6 SIGABRT)

--- command ---
PATH='/root/systemd/build:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin' SYSTEMD_LANGUAGE_FALLBACK_MAP='/root/systemd/src/locale/language-fallback-map' SYSTEMD_KBD_MODEL_MAP='/root/systemd/src/locale/kbd-model-map' /root/systemd/build/test-cgroup
--- stderr ---
Found container virtualization none.
/* test_cg_split_spec */
/* test_cg_create */
Found cgroup on /sys/fs/cgroup/systemd, legacy hierarchy
Paths for test:
/test-a
/test-b
Assertion 'streq(path, test_b)' failed at src/test/test-cgroup.c:79, function test_cg_create(). Aborting.
```

```
+ timeout --foreground 600 env --unset=UNIFIED_CGROUP_HIERARCHY --unset=SYSTEMD_NSPAWN_UNIFIED_HIERARCHY /root/systemd/build/systemd-nspawn -U --private-network --machine=TEST-01-BASIC --register=no --kill-signal=SIGKILL --directory=/var/tmp/systemd-test-TEST-01-BASIC/unprivileged-nspawn-root --setenv=SYSTEMD_UNIT_PATH=/usr/lib/systemd/tests/testdata/testsuite-01.units:/usr/lib/systemd/tests/testdata/units: /usr/lib/systemd/systemd root=/dev/sda1 rw raid=noautodetect rd.luks=0 loglevel=2 init=/usr/lib/systemd/systemd console=ttyS0 selinux=0 SYSTEMD_UNIT_PATH=/usr/lib/systemd/tests/testdata/testsuite-01.units:/usr/lib/systemd/tests/testdata/units: systemd.unit=testsuite.target systemd.wants=testsuite-01.service systemd.wants=end.service user_namespace.enable=1 systemd.unit=testsuite.target systemd.wants=testsuite-01.service systemd.wants=end.service
Spawning container TEST-01-BASIC on /var/tmp/systemd-test-TEST-01-BASIC/unprivileged-nspawn-root.
Press ^] three times within 1s to kill container.
Selected user namespace base 590217216 and range 65536.
systemd 246-25-g9f168e8 running in system mode. (+PAM +AUDIT +SELINUX +IMA -APPARMOR +SMACK +SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ +LZ4 -ZSTD +SECCOMP +BLKID +ELFUTILS +KMOD -IDN2 +IDN +PCRE2 default-hierarchy=unified)
Detected virtualization systemd-nspawn.
Detected architecture x86-64.

Welcome to CentOS Linux 7 (Core)!

Set hostname to <systemd-testsuite>.
Initializing machine ID from container UUID.
sd-netlink: Failed to enable NETLINK_EXT_ACK option, ignoring: Protocol not available
Failed to add address 127.0.0.1 to loopback interface: File exists
Failed to add address ::1 to loopback interface: File exists
Successfully brought loopback interface up
Failed to read AF_UNIX datagram queue length, ignoring: No such file or directory
Setting 'fs/file-max' to '9223372036854775807'.
Failed to bump fs.file-max, ignoring: Permission denied
Setting 'fs/nr_open' to '2147483640'.
Failed to bump fs.nr_open, ignoring: Permission denied
Found cgroup on /sys/fs/cgroup/systemd, legacy hierarchy
Using cgroup controller name=systemd. File system hierarchy is at /sys/fs/cgroup/systemd.
Failed to create /init.scope control group: Permission denied
Failed to allocate manager object: Permission denied
[[0;1;31m!!!!!![0m] Failed to allocate manager object.
Exiting PID 1...
Container TEST-01-BASIC failed with error code 255.
```

The original results can be found here: https://ci.centos.org/job/systemd-pr-build/13272/artifact//systemd-centos-ci/index.html
I rescheduled the job just to see if isn't caused by a broken machine, but the currently running job is failing in the same manner.

keszybz commented:
Yep, condition was inverted in one place.

vcaputo commented:
What is the impetus for this?  Can you provide a description of the intention behind this PR instead of "This came out longer than I expected..."?

Maybe this fell out of some other work/conversations, but it's unclear in isolation here.

keszybz commented:
Those helpers were internally marked as "smelly, should not be used", and we've been slowly removing any uses. @poettering's comment in https://github.com/systemd/systemd/pull/16590#discussion_r461447281 was the latest motivation.

vcaputo commented:
Thanks, so is this purely house cleaning then?

Is there an explanation somewhere on the rationale for the smelly classification?

keszybz commented:
> Thanks, so is this purely house cleaning then?

Mostly. There are some minor functional changes as noted in individual commits.

> Is there an explanation somewhere on the rationale for the smelly classification?

Lack of error handling.

keszybz commented:
When reviewing #16321 I realized that strv_split_extract() and strv_split_full() do almost exactly the same thing, and the latter should be removed. So the last few commits are rewritten to get rid of strv_split_full() and then rename the old strv_split_extract() to strv_split_full().

I also added some justification to the commits to make @vcaputo happy :)

poettering commented on src/libsystemd/sd-login/sd-login.c:
```diff
@@ -331,35 +333,29 @@ static int file_of_seat(const char *seat, char **_p) {
+        xsprintf(t, UID_FMT, uid);
 
-        return 0;
+        return string_contains_word(content, ",", t);
```
Hmm, this needs `EXTRACT_DONT_COALESCE_SEPARATORS` I guess.

keszybz replied:
This would only matter if the word in question was "". I guess I can rework it like that, but I think any difference is theoretical. (Does the string `" "` contain `""`? Does `""`? Does `"  "`?)

keszybz replied:
I made it use EXTRACT_DONT_COALESCE_SEPARATORS if separators != NULL. This seems to cover the intent correctly without having an explicit parameter.

poettering commented on src/basic/string-util.c:
```diff
@@ -1219,7 +1220,12 @@ int string_contains_word(const char *string, const char *separators, const char
                         return r;
                 if (r == 0)
                         return false;
```
given our usual logic to initialize all return params if we return >= 0 i think we should set `*ret_word = NULL;` here in the `false` branch

poettering commented on src/shared/sleep-config.c:
```diff
@@ -113,34 +112,27 @@ int can_sleep_state(char **types) {
+                return log_debug_errno(r, "Failed to parse /sys/power/state: %m");
+        if (r > 0)
+                log_debug("Sleep mode \"%s\" is supported by the kernel.", found);
+        else if (DEBUG_LOGGING) {
```
the previous branch only does debug logging too. Maybe encapsulate both this and the previous branch in a single `if (DEBUG_LOGGING)…`?

keszybz replied:
It comes out longer, and we need to have  a double set of {}... I think it's better to leave as-is.

poettering commented on src/login/logind-inhibit.c:
```diff
@@ -485,29 +485,34 @@ const char *inhibit_what_to_string(InhibitWhat w) {
+                _cleanup_free_ char *word = NULL;
+                int r;
+
+                r = extract_first_word(&p, &word, ":", 0);
```
EXTRACT_DONT_COALESCE?

poettering commented on src/basic/cgroup-util.c:
```diff
@@ -718,27 +712,25 @@ int cg_pid_get_path(const char *controller, pid_t pid, char **path) {
-                                }
-                        if (!found)
+
+                        r = string_contains_word(l, ",", controller_str);
```
EXTRACT_DONT_COALESCE_SEPARATORS?

poettering commented on src/libsystemd/sd-login/sd-login.c:
```diff
@@ -917,17 +916,25 @@ _public_ int sd_machine_get_ifindices(const char *machine, int **ret_ifindices)
         for (size_t i = 0; tt[i]; i++) {
-                ifindices[n] = parse_ifindex(tt[i]);
-                if (ifindices[n] > 0)
+                int ind = parse_ifindex(tt[i]);
```
we should probably propagate parse errors here...

poettering commented on src/cryptsetup/cryptsetup.c:
```diff
@@ -288,19 +288,19 @@ static int parse_one_option(const char *option) {
+                _cleanup_free_ char *word = NULL;
+                int r;
+
+                r = extract_first_word(&options, &word, NULL, 0);
```
extract_dont_coalesce_sep

poettering commented on src/delta/delta.c:
```diff
@@ -545,28 +545,33 @@ static int help(void) {
 
-        FOREACH_WORD_SEPARATOR(word, l, flag_str, ",", state) {
-                if (strneq("masked", word, l))
+                r = extract_first_word(&flag_str, &word, NULL, 0);
```
separator changed from "," to whitespace?

poettering commented on src/journal/sd-journal.c:
```diff
@@ -951,74 +951,70 @@ _public_ int sd_journal_get_cursor(sd_journal *j, char **cursor) {
 
-                if (l < 2 || word[1] != '=')
-                        return -EINVAL;
+                r = extract_first_word(&p, &word, ";", 0);
```
extract_dont_coalesce

keszybz commented:
Updated.

poettering commented:
hmmm needs rebase

keszybz commented:
Rebased to fix the trivial conflict, no other changes.

poettering commented on src/test/test-string-util.c:
```diff
@@ -364,22 +364,28 @@ static void test_strcmp_ptr(void) {
 
-        FOREACH_WORD(word, l, test, state)
-                assert_se(strneq(expected[i++], word, l));
+        int i = 0, r;
```
doesn't matter at all, but i would ideally be size_t here...

poettering commented:
also needs a rebase now

keszybz commented on src/basic/string-util.c:
```diff
@@ -1209,11 +1210,13 @@ int string_extract_line(const char *s, size_t i, char **ret) {
          * don't coalesce separators. */
         const ExtractFlags flags = separators ? EXTRACT_DONT_COALESCE_SEPARATORS : 0;
 
+        const char *found = NULL;
```
To make it clear that the comment applies to the `flags` param.

keszybz commented on src/login/logind-inhibit.c:
```diff
@@ -485,29 +485,34 @@ const char *inhibit_what_to_string(InhibitWhat w) {
+
+                r = extract_first_word(&p, &word, ":", EXTRACT_DONT_COALESCE_SEPARATORS);
+                if (r < 0)
+                        return r;
```
I added an assert_cc that the value is fine. I think this is good enough — the compiler deciding to make the type smaller than a word would be very strange.

keszybz replied:
Well sure, if it decides, the code will fail to compile. If you absolutely require this, I'll change it, but our code is very unlikely to be compiled by anything except gcc or clang (because there's just too many extensions that are not supported elsewhere), and both gcc and clang are very unlikely to start doing something different here.

poettering replied:
I still think this should be changed. I wonder if coverity or so would be tripped up by this the instant this is merged...

Also not that the choice of word size might depend on arch, and the casts here are implicit, i.e. the compiler might assume the user has checked the value fits into the type, and weird stuff might happen if that's not the case...

keszybz replied:
I changed the return type to `int` now.

keszybz commented on src/libsystemd/sd-login/sd-login.c:
```diff
@@ -893,47 +893,46 @@ _public_ int sd_machine_get_class(const char *machine, char **class) {
-                }
+                ind = parse_ifindex(tt[i]);
+                if (ind < 0)
+                        return -EUCLEAN;
```
Because EINVAL would be confused as "the parameters passed to the function are bad". I wanted to distinguish that case. I changed that to `return ind == -EINVAL ? -EUCLEAN : ind;` now.

keszybz commented on src/libsystemd/sd-login/sd-login.c:
```diff
@@ -893,47 +893,46 @@ _public_ int sd_machine_get_class(const char *machine, char **class) {
-                ni[nr++] = ifi;
+                ifindices[n++] = ind;
         }
 
```
I converted sd_seat_get_sessions() above like that, but here we don't know the size of the array. So we'd need to reallocate on the fly or parse twice. So I think this approach is OK. The number of indices is likely to be quite small, so I don't think it matters much either way.

keszybz commented:
Rebased and updated as suggested, except where I replied above.

keszybz commented:
Oops, an unused variable. I'll update in a sec.

poettering commented on src/libsystemd/sd-login/sd-login.c:
```diff
@@ -929,10 +931,13 @@ _public_ int sd_machine_get_ifindices(const char *machine, int **ret_ifindices)
                         return ind == -EINVAL ? -EUCLEAN : ind;
 
-                ifindices[n++] = ind;
+                if (ifindices)
```
you probably want to check ret_ifindices here (or you need to initialize ifindices to NULL up top when ret_ifindices is NULL)

keszybz replied:
Good catch.

poettering commented on src/journal/sd-journal.c:
```diff
@@ -951,74 +951,69 @@ _public_ int sd_journal_get_cursor(sd_journal *j, char **cursor) {
-                item = strndup(word, l);
-                if (!item)
-                        return -ENOMEM;
+                if (!word[0] || word[1] != '=')
```
given that these are chars, and not really bools, i'd do the check explicitly here: `word[0] == 0 || …`

poettering commented on src/shared/conf-parser.h:
```diff
@@ -255,29 +255,30 @@ typedef enum Disabled {
-                                return log_oom();                              \
+                        r = extract_first_word(&p, &en, NULL, 0);              \
+                        if (r < 0)                                             \
+                                return log_error_errno(r, "Failed to parse %s: %m", lvalue); \
```
log_syntax()

keszybz commented:
Updated. I changed the return type from InhibitWhat to int, but I didn't make the enum anonymous. (I think it's nicer to have it non-anonymous when debugging, and I don't see much difference otherwise.)
